File: debuglog-annotations/src/main/java/it/sephiroth/android/library/debuglog/DebugLogger.java
package it.sephiroth.android.library.debuglog;

import android.util.Log;

/**
 * Created by quinn on 2019/1/8
 */
@SuppressWarnings("unused")
public class DebugLogger {
    public static final DebugLogHandler NullLogger = new DebugLogHandler();
    public static final DebugLogHandler DefaultLogger;

    static {
        DefaultLogger = new DebugLogHandler() {

            @Override
            protected void logEnter(int priority, String tag, String methodName, String params) {
                Log.println(priority, tag, String.format("\u21E2 %s[%s]", methodName, params)); // ⇢
            }

            @Override
            protected void logExit(int priority, String tag, String methodName, long costedMillis, String result) {
                Log.println(priority, tag, String.format("\u21E0 %s[%sms] = %s", methodName, costedMillis, result)); // ⇠
            }
        };
    }

    static DebugLogHandler DEFAULT_IMPL = DefaultLogger;

    public static void installLog(DebugLogHandler loggerHandler) {
        DEFAULT_IMPL = loggerHandler;
    }

}

File: debuglog-annotations/src/main/java/it/sephiroth/android/library/debuglog/DebugLogger.kt
package it.sephiroth.android.library.debuglog

import android.util.Log

object DebugLogger {
    @JvmField
    val nullLogger = DebugLogHandler()

    @JvmField
    val defaultLogger: DebugLogHandler? = object : DebugLogHandler() {
        override fun logEnter(priority: Int, tag: String, methodName: String, params: String?) {
            // ⇢
            Log.println(priority, tag, String.format("\u21E2 %s[%s]", methodName, params))
        }

        override fun logExit(priority: Int, tag: String, methodName: String, costedMillis: Long, result: String?) {
            // ⇠
            if (null != result) {
                Log.println(priority, tag, String.format("\u21E0 %s[%sms] = %s", methodName, costedMillis, result))
            } else {
                Log.println(priority, tag, String.format("\u21E0 %s[%sms]", methodName, costedMillis))
            }
        }
    }

    @JvmField
    var DEFAULT_IMPL = defaultLogger

    /**
     * Install a new logger
     */
    fun installLog(loggerHandler: DebugLogHandler?) {
        DEFAULT_IMPL = loggerHandler
    }

}
